### PR TITLE
KAFKA-16000 Migrated MembershipManagerImplTest away from ConsumerTestBuilder

### PR DESCRIPTION
**Pull request by brenden20**
I completely migrated MembershipManagerImplTest away from ConsumerTestBuilder and removed all spy objects. All tests passing, none were removed.

### CODE REVIEW

**Comment by brenden20** [comment]
@philipnee @lianetm @kirktrue I am getting a checkstyle failure, I have one too many class data abstractions, I am unsure on where I can fix this. Any ideas?

**Comment by lianetm** [comment]
Hey @brenden20 , assuming it's something we cannot avoid in this case, you can suppress it with @SuppressWarnings("ClassDataAbstractionCoupling") (it's done in other places too)

**Comment by brenden20** [comment]
> Hey @brenden20 , assuming it's something we cannot avoid in this case, you can suppress it with @SuppressWarnings("ClassDataAbstractionCoupling") (it's done in other places too)

Great to know! I will take another look at the code to make sure it is necessary, if so I will add that annotation. Thank you!

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: we shouldn't need an actual `commitRequestManager` here, where we don't need anything specific related to commits and a mock should do. I guess you could remove it and just return a completed future for the autoCommit before revocation :

`when(commitRequestManager.maybeAutoCommitSyncBeforeRevocation(anyLong())).thenReturn(CompletableFuture.completedFuture(null));`

Probably doing it once inside the `createMemberInStableState` will allow you to remove most of the createCommitRequestManager calls

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: Even though I totally agree with removing all the spies there were on other components, I would lean towards spying only on this membershipManager component in cases like this, where it seems valuable to verify on it. It would only mean that we spy on it when needed with `membershipManager = spy(createMemberInStableState())` on the specific tests that do verify, and not spying on membershipmanager for all test that in the end don;t need it. Makes sense?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: uhm we shouldn't call this here ourselves. The test wants to ensure that when the callbacks complete, the membershipMgr internally calls this `transitionToSendingLeaveGroup`. If for some reason the transition to leaving was not happening we should review what's missing/failing that does not let the flow make it to [this](https://github.com/apache/kafka/blob/f2a552a1ebaa0ce933f90343655b63e2472856d9/clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java#L695) point, where the internal call to transitionToSendingLeaveGroup should happen.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: similar to comment above, I would keep this kind of verification (spying on the membershipMgr only). The membership mgr holds the state machine for the member as part of a group, it's all intertwined, we've seen how small changes in one membership func affects/breaks states and transitions, so all coverage we can maintain here is important.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: let's remove the comment if we don't need the line anymore

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: heads-up, having these as mocks means we cannot retrieve the events like we used to (backgroundEventQueue.poll), that I notice is still used to get the event to `performCallback`. I would say having it as a mock is the right thing to do, but we need to fix how we get the events from it. Ex. [this](https://github.com/apache/kafka/blob/53ec9733c2bf8b122fbd193d8d00b8e25cbd648f/clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java#L2165-L2169) section to get the callback event to complete should be updated to retrieve the event with an argument captor instead (similar to how it's done [here](https://github.com/apache/kafka/blob/f2a552a1ebaa0ce933f90343655b63e2472856d9/clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java#L807-L809), but retrieving a `ConsumerRebalanceListenerCallbackNeededEvent`)

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: I would aim for not having an actual commitRequestManager here, a mock should do, but I may be missing something. Check my other comments to get rid of the `createCommitRequestManager` (and maybe we can end up removing this?)

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: uhm removing this will make that in all tests that want to block on commit we will have to call the same `when(commitRequestManager.maybeAutoCommitSyncBeforeRevocation...` we had here I guess? I do see that you had to add this line in many places in the end. Again, seems related to how we're using an insntace of commitRequestManager mentioned in comments above. (and if we are able to move to a mocked commitManager this should probably stay?)

**Comment by lianetm** [comment]
Hey @brenden20, thanks for the changes! Took a first pass, left a few comments. I see some other calls around tests blocked on commits that seem confusing but I expect they may all align better if we review how we use the commit manager and are able to simplify to a mock maybe. Thanks!

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (brenden20) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: Makes sense, I have reverted that change

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (brenden20) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: Removed now

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (brenden20) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: I have reverted changes here

**Comment by brenden20** [comment]
@lianetm thank you for the feedback! I have implemented all suggestions, it is looking really good now I think!

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: let's avoid this new line changes that do not seem relevant? (here and below)

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: what about `assertInstanceOf(ConsumerRebalanceListenerCallbackNeededEvent.class, backgroundEventQueue.peek());`

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: the action on if and else seem exactly the same, what's the intention?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: if(expectedTopicPartitionAssignment.isEmpty())

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: I wouldn't expect us to need to manually update the assignment here (because the empty assignment comes from the heartbeat data on ln 319 above). Could you check if/why we need it?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: This is a nice catch/improvement. The `assignFromSubscribedAwaitingCallback` is the one that the membershipMgr calls (and that one internally calls the `assignFromSubscribed`, good catch).

On another note, we know no partitions were added on this reconciliation so we could verify with empty collection as second param (instead of anyCollection()). And I expect you could then remove the eq on the 1st param too.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: similar to above comment, here the second param we know it should be the same set of partitions 1 and 2 for topic2 (exactly the same as the first param, because the assignment is expected to be [t2p1, t2p2], and the added partitions the same set, because the member did not own any of them) .
I notice several following calls to assignFromSubscribedAwaitingCallback that used to have the exact set they were expecting now have any(), could you check and add it back when possible?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: here we probably know what we expect as assignment in the first param instead of any() (is this maybe a mixup with the if/else verify right below?

**Comment by lianetm** [comment]
Hey @brenden20, I just completed another pass, great simplification with the latest updates! Left some comments. Thanks!

**Comment by brenden20** [comment]
Thank you for the feedback @lianetm, I just made all the suggested improvements!

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: is there a reason to use the `eq` here, that we're providing the to specific arguments?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: we could simplify removing this and just reference `Collections.emptySet()` on the 2nd param of the assignFromSubscribedAwaitingCallback

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (lianetm) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: if we don't need it let's remove it in other places when possible

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: hey why exactly do we need this?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java:1931]: the original implementation was better, any reason for changing that? `verify(backgroundEventHandler).add(any(ConsumerRebalanceListenerCallbackNeededEvent.class));`

it was better because backgroundEventQueue is in the implementation.  we just need to make sure the event was added to the handler.

Comment (brenden20) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java:1931]: Since ```backgroundEventHandler``` is an actual instance of an object, we cannot ```verify()``` on it. So in order to still test what it was trying to test, the two assert statements must be there instead

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (brenden20) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: When we ```verify(subscriptionState).assignFromSubscribedAwaitingCallback()```, we want to make sure that the assignments are being calculated and propagated properly. Having an expected set of partitions for the assignment makes the test more strict, yet still correct.

Comment (lianetm) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: @brenden20 I believe @philipnee 's point is about why we need the `expectedSet` and can't get away with `expectedAssignment` only? I think we could indeed remove expectedSet here (we do need it in other similar checks to sort the set or add elems, but since this one has a single elem we don't need to have it I expect)

Comment (brenden20) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: @lianetm I see now, I have made the change

**Review by lianetm — APPROVED** [review]
Thanks for the great improvement and all the updates @brenden20 ! Let's make sure @philipnee 's concerns are clarified but this LGTM. Thanks!

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java:1931]: does changing `backgroundEventHandler` to a mock break a lot of tests?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: isn't `Set<TopicPartition> expectedAssignment ` already a set?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (mjsax) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: For my own education: why move this here and not keep `logContext` as `final`?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (mjsax) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java:724]: Why do we verify something else here? This PR only updates test code, so where does this change in behavior come from? Seems I am missing something.

Comment (brenden20) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java:724]: ```assignFromSubscribed()``` is called from ```assignFromSubscribedAwaitingCallback()```. Since ```subscriptionState``` was changed from a spy to a mock, we could not test for this behavior. So instead I changed the verify to be the calling function ```assignFromSubscribedAwaitingCallback()``` and the correct, expected parameters. If ```assignFromSubscribedAwaitingCallback()``` is getting the correct parameters, then it should implicitly verify that  ```assignFromSubscribed()``` is being called with the correct parameter.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (brenden20) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: It was mostly to keep consistency with variables being private/private and final. Although I did not consider that I could make it ```private static final``` and it would still work without issue and keep the variables consistent. I am going to make that change now.

**Comment by brenden20** [comment]
@mjsax thank you for the feedback! I made a small change with ```logContext``` and replied to your other comment. If there is anything else let me know!

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: similar to the comment above, can we just verify against the expectedAssignment?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java:120]: fwiw: `new Mocktime()` is sufficient. just a comment.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (brenden20) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java:1931]: Yeah, breaks 16/81, all test failures are on asserting an empty queue or number of events in the queue

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (brenden20) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java]: Yes, fixed now

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java
Comment (brenden20) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImplTest.java:120]: Got it, I will keep that in mind

**Comment by brenden20** [comment]
All test failures are unrelated to the changes made in this PR

**Comment by mjsax** [comment]
Thanks for the PR @brenden20. Merged to `trunk`.

**Pull request merged**
